### DISCUSSION TITLE
PSBT API support

### DISCUSSION THREAD

**Discussion by DerpHerpenstein** [CounterpartyXCP/CIPs#131] [General]
CIP: XXX
        Title: PSBT API Support
        Author: Derp Herpenstein
        Discussions-To: ??
        Status: Draft
        Type: ??
        Created: 2024-1-12

# Abstract
Add the option to return transactions as PSBTs in the API

# Rationale
Counterparty is unnecessarily cumbersome for newcomers.  The ordinals ecosystem has shown that simple tooling is needed to lower the friction of end users.  Modern Bitcoin web based wallets are largely used as part of this onboarding process.  The ability to send bitcoin to a wallet like unisat where the user can then click a single button that says "buy" to purchase an asset cannot be understated.

These browser based wallets rely on PSBTs for the trading of ordinals, and as such, the ability to sign a PSBT is simple and straight forward.  The addition of a PSBT option in transaction generation will allow the large players in the ordinals ecosystem to seemlessly integrate counterparty transactions.

# Definitions
- `PSBT` - Partially signed bitcoin transaction

## Limitations 
Although this will allow for easy generation of transactions, it will not allow for PSBT based single transaction trading.  As a follow on to this, a CIP that updates the send function to have a mechanism for attaching assets to UTXOs will need to be created to deterministically use PSBTs to transfer assets in a single transaction. 

When these two CIPs are complete, all counterparty assets will be able to be traded with BTC and/or ordinals in a single transaction and user can easily be onboard with the same tools ordinals are currently using.

# Implementation
In counterparty-lib in the file lib/transactions.py an additional function called constructPsbt will be created.  It will take the same parameters as the construct function.  This function will call construct and with the result it will use the bitcoin backend to call the function converttopsbt to generate the hex. Then as a safety measure, the psbt will be deconstructed to verify the parameters match those expected before returning the psbt.

# Copyright
This document is placed in the public domain.

**Comment by DerpHerpenstein** [comment]
For more information about PSBT trading, see [this discussion](https://github.com/CounterpartyXCP/cips/discussions/104)

**Comment by adamkrellenstein** [comment]
I think this is a great idea!
